### PR TITLE
Campaigns controller part 2

### LINKED ISSUE #98: CampaignController

### PR DESCRIPTION
**Pull request by ganhongyao**
Closes #98

sample POST

```json
{
    "name": "siva",
    "description": "testing1234",
    "promisedAmount": 1000,
    "start": "2022-10-10T14:48:00.000Z",
    "end": "2022-10-11T14:48:00.000Z",
    "primaryDonor": {
        "name": "Siva",
        "email": "siva@siva.com"
    },
    "couponDenomination": 10,
    "charities": [
        {
            "charity": {
                "id": 3
            },
            "givingSgUrl": "https://givingsg.com/1"
        },
        {
            "charity": {
                "id": 2
            },
            "givingSgUrl": "https://givingsg.com/2"
        }
    ],
    "imageBase64": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAApgAAAKYB3X3/OAAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAANCSURBVEiJtZZPbBtFFMZ/M7ubXdtdb1xSFyeilBapySVU8h8OoFaooFSqiihIVIpQBKci6KEg9Q6H9kovIHoCIVQJJCKE1ENFjnAgcaSGC6rEnxBwA04Tx43t2FnvDAfjkNibxgHxnWb2e/u992bee7tCa00YFsffekFY+nUzFtjW0LrvjRXrCDIAaPLlW0nHL0SsZtVoaF98mLrx3pdhOqLtYPHChahZcYYO7KvPFxvRl5XPp1sN3adWiD1ZAqD6XYK1b/dvE5IWryTt2udLFedwc1+9kLp+vbbpoDh+6TklxBeAi9TL0taeWpdmZzQDry0AcO+jQ12RyohqqoYoo8RDwJrU+qXkjWtfi8Xxt58BdQuwQs9qC/afLwCw8tnQbqYAPsgxE1S6F3EAIXux2oQFKm0ihMsOF71dHYx+f3NND68ghCu1YIoePPQN1pGRABkJ6Bus96CutRZMydTl+TvuiRW1m3n0eDl0vRPcEysqdXn+jsQPsrHMquGeXEaY4Yk4wxWcY5V/9scqOMOVUFthatyTy8QyqwZ+kDURKoMWxNKr2EeqVKcTNOajqKoBgOE28U4tdQl5p5bwCw7BWquaZSzAPlwjlithJtp3pTImSqQRrb2Z8PHGigD4RZuNX6JYj6wj7O4TFLbCO/Mn/m8R+h6rYSUb3ekokRY6f/YukArN979jcW+V/S8g0eT/N3VN3kTqWbQ428m9/8k0P/1aIhF36PccEl6EhOcAUCrXKZXXWS3XKd2vc/TRBG9O5ELC17MmWubD2nKhUKZa26Ba2+D3P+4/MNCFwg59oWVeYhkzgN/JDR8deKBoD7Y+ljEjGZ0sosXVTvbc6RHirr2reNy1OXd6pJsQ+gqjk8VWFYmHrwBzW/n+uMPFiRwHB2I7ih8ciHFxIkd/3Omk5tCDV1t+2nNu5sxxpDFNx+huNhVT3/zMDz8usXC3ddaHBj1GHj/As08fwTS7Kt1HBTmyN29vdwAw+/wbwLVOJ3uAD1wi/dUH7Qei66PfyuRj4Ik9is+hglfbkbfR3cnZm7chlUWLdwmprtCohX4HUtlOcQjLYCu+fzGJH2QRKvP3UNz8bWk1qMxjGTOMThZ3kvgLI5AzFfo379UAAAAASUVORK5CYII="
}
```

### CODE REVIEW

File: backend/app/controllers/campaigns_controller.rb [hunk 1]
```diff
@@ -1,15 +1,12 @@
 # frozen_string_literal: true
 
+# rubocop:disable Metrics/ClassLength
 class CampaignsController < ApplicationController
```
Comment (sivayogasubramanian) [inline backend/app/controllers/campaigns_controller.rb:4]: For all methods here, we should avoid using `params`. Let's use `campaign_params` so that unwanted params are caught.

File: backend/app/controllers/campaigns_controller.rb [hunk 2]
```diff
@@ -48,9 +63,92 @@ def set_campaign
...
+  def create_campaign_charities
+    charity_params = params[:charities]
+
+    @campaign.campaign_charities = charity_params.map do |campaign_charity_param|
+      CampaignCharity.new(charity_id: campaign_charity_param[:charity][:id],
+                          giving_sg_url: campaign_charity_param[:giving_sg_url])
+    end
+  end
+
+  def create_or_update_campaign_charities
```
Comment (sivayogasubramanian) [inline backend/app/controllers/campaigns_controller.rb:90]: Where does this method update campaign_charities?

File: backend/app/controllers/campaigns_controller.rb [hunk 3]
```diff
@@ -48,9 +63,92 @@ def set_campaign
...
+
+  def end_params?
+    end_params = params[:end]
+    end_params.present? && end_params['from'].present? && end_params['to'].present?
+  end
+
+  def scoped_with_dates
+    start_params = params[:start]
+    starts_before = start_params['to'] if start_params.present?
+    starts_after = start_params['from'] if start_params.present?
```
Comment (sivayogasubramanian) [inline backend/app/controllers/campaigns_controller.rb:135]: Doesn't start_params return a boolean itself?

File: backend/app/controllers/campaigns_controller.rb [hunk 4]
```diff
@@ -48,9 +63,92 @@ def set_campaign
...
+
+    end_params = params[:end]
+    ends_before = end_params['to'] if end_params.present?
+    ends_after = end_params['from'] if end_params.present?
```
Comment (sivayogasubramanian) [inline backend/app/controllers/campaigns_controller.rb:139]: Same here for end_params.

File: backend/app/controllers/campaigns_controller.rb [hunk 5]
```diff
@@ -48,9 +63,92 @@ def set_campaign
...
+
+    Campaign.starts_before(starts_before).starts_after(starts_after)
+            .ends_before(ends_before).ends_after(ends_after)
+  end
+
+  def scoped_with_status
```
Comment (sivayogasubramanian) [inline backend/app/controllers/campaigns_controller.rb:145]: Can we simplify it to 

```
status_params[:is_active].present? && status_params[:is_active] == 'true'
```

Also is the type a string? If its not 

```
status_params[:is_active].present? && status_params[:is_active]
```

File: backend/app/controllers/charities_controller.rb [hunk 6]
```diff
@@ -4,37 +4,35 @@ class CharitiesController < ApplicationController
...
   def show; end
 
   def create
-    @charity = Charity.new(charity_params.except(:logo_base64, :image_base64))
-    @charity.logo.attach(data: charity_params[:logo_base64]) unless charity_params[:logo_base64].nil?
-    @charity.image.attach(data: charity_params[:image_base64]) unless charity_params[:image_base64].nil?
+    @charity = Charity.new(charity_params)
+    @charity.logo.attach(data: params[:logo_base64]) unless image_params[:logo_base64].nil?
+    @charity.image.attach(data: params[:image_base64]) unless image_params[:image_base64].nil?
     @charity.save!
```
Comment (sivayogasubramanian) [inline backend/app/controllers/charities_controller.rb:17]: Same here. Let's use `image_params` instead of params.

Also help to flip the logic as per CampaignsController.

File: backend/app/controllers/charities_controller.rb [hunk 7]
```diff
@@ -4,37 +4,35 @@ class CharitiesController < ApplicationController
...
 
     add_success_message "Charity, \"#{@charity.name}\", successfully created!"
     render :show, status: :created
   end
 
   def update
-    @charity.update(charity_params.except(:logo_base64, :image_base64))
+    @charity.update(charity_params)
```
Comment (sivayogasubramanian) [inline backend/app/controllers/charities_controller.rb:24]: Can help to make this a transaction as per CharitiesController?

File: backend/app/controllers/charities_controller.rb [hunk 8]
```diff
@@ -55,6 +53,10 @@ def set_charity
   end
 
   def charity_params
-    params.require(:charity).permit(:name, :description, :website_url, :logo_base64, :image_base64)
+    params.require(:charity).permit(:name, :description, :website_url)
+  end
+
+  def image_params
+    params.permit(:logo_base64, :image_base64)
   end
```
Comment (sivayogasubramanian) [inline backend/app/controllers/charities_controller.rb:61]: Shouldn't this be inside :charity?

File: backend/app/controllers/interests_controller.rb [hunk 9]
```diff
@@ -4,24 +4,25 @@ class InterestsController < ApplicationController
...
   def index
     @interests = Interest.all
   end
 
   def show; end
 
   def create
-    @interest = Interest.create!(interest_params)
+    @interest = Interest.new(interest_params)
+    @interest.charity_ids = params[:charity_ids]
```
Comment (sivayogasubramanian) [inline backend/app/controllers/interests_controller.rb:15]: Same comment about params.

File: backend/app/controllers/secondary_donations_controller.rb [hunk 10]
```diff
@@ -2,14 +2,15 @@
...
 
   def index
     @secondary_donations = SecondaryDonation.all
   end
 
   def create
-    @secondary_donation = SecondaryDonation.create!(secondary_donation_params)
+    @secondary_donation = SecondaryDonation.new(secondary_donation_params)
+    @secondary_donation.campaign_charity = CampaignCharity.find(params[:campaign_charity_id])
+    @secondary_donation.save!
```
Comment (sivayogasubramanian) [inline backend/app/controllers/secondary_donations_controller.rb:13]: I see that you are manually linking stuff.

Question: Why is this needed? Doesn't rails know to link it when there is a `campaign_charity_id` present?

**Review by sivayogasubramanian — APPROVED** [review]
Not testing as it will be easier to test when the frontend is up.

File: backend/app/controllers/campaigns_controller.rb (hunk 3 above)
Comment (ganhongyao) [reply backend/app/controllers/campaigns_controller.rb:135]: That's `start_params?` though, this is `start_params`.

File: backend/app/controllers/campaigns_controller.rb (hunk 5 above)
Comment (ganhongyao) [reply backend/app/controllers/campaigns_controller.rb:145]: Yes it is a string

File: backend/app/controllers/charities_controller.rb (hunk 8 above)
Comment (ganhongyao) [reply backend/app/controllers/charities_controller.rb:61]: I removed the manual `wrap_parameters` method (with the help of @ianyong LOL), so now we don't manually wrap the JSON params. This means that the wrapping is done by Rails, who will only wrap the keys found in `Model.attribute_names`. `logo_base64` and `image_base64` are not attributes, so they are not found inside the wrapped `:charity` .

File: backend/app/controllers/secondary_donations_controller.rb (hunk 10 above)
Comment (ganhongyao) [reply backend/app/controllers/secondary_donations_controller.rb:13]: Same as the above comment - `campaign_charity_id` is not an attribute name thus we have to link the associations manually.

File: backend/app/controllers/campaigns_controller.rb (hunk 2 above)
Comment (ganhongyao) [reply backend/app/controllers/campaigns_controller.rb:90]: Done
